File: backend/app/services/llm_service.py
from __future__ import annotations

import logging

from app.services.git_features import GitFeatures
from app.services.scorer import ScoredCandidate

logger = logging.getLogger(__name__)


def _format_comments(issue) -> str:
    """Return up to 3 comment snippets (200 chars each) for the LLM prompt."""
    snippets = []
    for c in issue.jira_comments[:3]:
        text = (c.body_digest or "").strip()
        if text:
            snippets.append(f"- {text[:200]}")
    return "\n".join(snippets)


def _make_client():
    from openai import OpenAI
    import os
    return OpenAI(api_key=os.environ["OPENAI_API_KEY"])


def rerank_candidates(
    git_features: GitFeatures,
    scored: list[ScoredCandidate],
    model: str,
    top_n: int = 10,
) -> list[ScoredCandidate]:
    if not scored:
        return scored

    candidates_to_rank = scored[:top_n]

    commit_text = "\n".join(git_features.commit_messages[:3])
    candidate_lines = "\n".join(
        f"{i + 1}. [{c.issue.issue_key}] {c.issue.summary} (score: {c.total_score:.0f})"
        for i, c in enumerate(candidates_to_rank)
    )

    prompt = (
        f"You are ranking Jira issues by relevance to a code change.\n\n"
        f"Commit messages:\n{commit_text}\n\n"
        f"Candidates (already pre-scored):\n{candidate_lines}\n\n"
        f"Return ONLY a comma-separated list of the candidate numbers in order of relevance, "
        f"most relevant first. Example: 3,1,2,5,4"
    )

    try:
        client = _make_client()
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=64,
            temperature=0,
        )
        order_str = response.choices[0].message.content.strip()
        indices = [int(x.strip()) - 1 for x in order_str.split(",") if x.strip().isdigit()]
        reranked = [candidates_to_rank[i] for i in indices if 0 <= i < len(candidates_to_rank)]
        # Append any candidates not mentioned by the LLM at the end
        mentioned = set(indices)
        for i, c in enumerate(candidates_to_rank):
            if i not in mentioned:
                reranked.append(c)
        return reranked + scored[top_n:]
    except Exception as exc:
        logger.warning("LLM rerank failed, using deterministic order: %s", exc)
        return scored


def generate_explanation(
    git_features: GitFeatures,
    top_candidates: list[ScoredCandidate],
    output_mode: str,
    model: str,
) -> str:
    if not top_candidates and output_mode not in ("git_only", "no_match"):
        return "No matching Jira issue found for this code change."

    commit_text = "\n".join(git_features.commit_messages[:3])

    if output_mode == "git_only":
        prompt = (
            f"Explain in 2-3 sentences why this code likely changed, "
            f"based only on the commit messages below. No Jira context is available.\n\n"
            f"Commits:\n{commit_text}"
        )
    elif output_mode == "combined":
        issues_text = "\n".join(
            f"- [{c.issue.issue_key}] {c.issue.summary}" for c in top_candidates[:3]
        )
        prompt = (
            f"Explain in 2-4 sentences why the selected code block changed. "
            f"Multiple Jira issues contributed to this change — describe the combined context.\n\n"
            f"Commit messages:\n{commit_text}\n\n"
            f"Contributing Jira issues:\n{issues_text}"
        )
    elif output_mode == "no_match":
        return "No matching Jira issue found for this code change."
    else:
        # single_issue (default)
        best = top_candidates[0]
        comments_text = _format_comments(best.issue)
        prompt = (
            f"Explain in 2-3 sentences why this code changed, "
            f"linking it to the Jira issue below.\n\n"
            f"Commit: {commit_text}\n"
            f"Jira issue [{best.issue.issue_key}]: {best.issue.summary}\n"
            f"Description: {best.issue.description_digest or '(none)'}"
            + (f"\nTop comments:\n{comments_text}" if comments_text else "")
        )

    try:
        client = _make_client()
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=256,
            temperature=0.3,
        )
        return response.choices[0].message.content.strip()
    except Exception as exc:
        logger.warning("LLM explanation failed: %s", exc)
        # Fallback: return a deterministic summary without LLM
        if top_candidates:
            best = top_candidates[0]
            return (
                f"This code was likely changed in relation to [{best.issue.issue_key}]: "
                f"{best.issue.summary}. (LLM explanation unavailable.)"
            )
        return f"Change context: {commit_text[:200]}"

File: plugin/bin/main/com/whyline/plugin/analyzer/LlmService.kt
package com.whyline.plugin.analyzer

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.intellij.openapi.diagnostic.Logger
import java.net.HttpURLConnection
import java.net.URL

private val LOG = Logger.getInstance("WhyLine.LlmService")
private val GSON = Gson()

object LlmService {

    fun rerankCandidates(
        features: GitFeatures,
        scored: List<ScoredCandidate>,
        model: String,
        apiKey: String,
        topN: Int = 10,
    ): List<ScoredCandidate> {
        if (scored.isEmpty()) return scored
        val toRank = scored.take(topN)
        val commitText = features.commitMessages.take(3).joinToString("\n")
        val candidateLines = toRank.mapIndexed { i, c ->
            "${i + 1}. [${c.issue.issueKey}] ${c.issue.summary} (score: ${"%.0f".format(c.totalScore)})"
        }.joinToString("\n")

        val prompt = "You are ranking Jira issues by relevance to a code change.\n\n" +
            "Commit messages:\n$commitText\n\n" +
            "Candidates (already pre-scored):\n$candidateLines\n\n" +
            "Return ONLY a comma-separated list of the candidate numbers in order of relevance, " +
            "most relevant first. Example: 3,1,2,5,4"

        return try {
            val response = chatCompletion(prompt, model, maxTokens = 64, temperature = 0.0, apiKey = apiKey)
                ?: return scored
            val indices = response.split(",")
                .mapNotNull { it.trim().toIntOrNull()?.minus(1) }
                .filter { it in toRank.indices }
            val reranked = indices.map { toRank[it] }.toMutableList()
            val mentioned = indices.toSet()
            for (i in toRank.indices) { if (i !in mentioned) reranked.add(toRank[i]) }
            reranked + scored.drop(topN)
        } catch (e: Exception) {
            LOG.warn("LLM rerank failed: ${e.message}")
            scored
        }
    }

    fun generateExplanation(
        features: GitFeatures,
        topCandidates: List<ScoredCandidate>,
        outputMode: String,
        model: String,
        apiKey: String,
    ): String {
        val commitText = features.commitMessages.take(3).joinToString("\n")

        val prompt = when (outputMode) {
            "git_only" ->
                "Explain in 2-3 sentences why this code likely changed, " +
                    "based only on the commit messages below. No Jira context is available.\n\nCommits:\n$commitText"

            "combined" -> {
                val issuesText = topCandidates.take(3).joinToString("\n") {
                    "- [${it.issue.issueKey}] ${it.issue.summary}"
                }
                "Explain in 2-4 sentences why the selected code block changed. " +
                    "Multiple Jira issues contributed to this change — describe the combined context.\n\n" +
                    "Commit messages:\n$commitText\n\nContributing Jira issues:\n$issuesText"
            }

            "no_match" -> return "No matching Jira issue found for this code change."

            else -> {
                if (topCandidates.isEmpty()) return "No matching Jira issue found for this code change."
                val best = topCandidates[0]
                val commentsText = best.issue.comments.take(3)
                    .mapNotNull { it.body?.trim()?.takeIf { b -> b.isNotEmpty() }?.take(200) }
                    .joinToString("\n") { "- $it" }
                "Explain in 2-3 sentences why this code changed, " +
                    "linking it to the Jira issue below.\n\n" +
                    "Commit: $commitText\n" +
                    "Jira issue [${best.issue.issueKey}]: ${best.issue.summary}\n" +
                    "Description: ${best.issue.description?.take(500) ?: "(none)"}" +
                    (if (commentsText.isNotBlank()) "\nTop comments:\n$commentsText" else "")
            }
        }

        return try {
            chatCompletion(prompt, model, maxTokens = 256, temperature = 0.3, apiKey = apiKey)
                ?: fallback(topCandidates, commitText)
        } catch (e: Exception) {
            LOG.warn("LLM explanation failed: ${e.message}")
            fallback(topCandidates, commitText)
        }
    }

    private fun fallback(candidates: List<ScoredCandidate>, commitText: String): String {
        if (candidates.isNotEmpty()) {
            val best = candidates[0]
            return "This code was likely changed in relation to [${best.issue.issueKey}]: ${best.issue.summary}. (LLM explanation unavailable.)"
        }
        return "Change context: ${commitText.take(200)}"
    }

    private fun chatCompletion(
        prompt: String,
        model: String,
        maxTokens: Int,
        temperature: Double,
        apiKey: String,
    ): String? {
        val body = GSON.toJson(mapOf(
            "model" to model,
            "messages" to listOf(mapOf("role" to "user", "content" to prompt)),
            "max_tokens" to maxTokens,
            "temperature" to temperature,
        ))
        return try {
            val json = httpPost("https://api.openai.com/v1/chat/completions", apiKey, body) ?: return null
            val obj = GSON.fromJson(json, JsonObject::class.java)
            obj.getAsJsonArray("choices")?.firstOrNull()?.asJsonObject
                ?.getAsJsonObject("message")?.get("content")?.asString?.trim()
        } catch (e: Exception) {
            LOG.warn("chatCompletion failed: ${e.message}")
            null
        }
    }

    private fun httpPost(url: String, apiKey: String, body: String): String? {
        val conn = URL(url).openConnection() as HttpURLConnection
        return try {
            conn.requestMethod = "POST"
            conn.doOutput = true
            conn.setRequestProperty("Authorization", "Bearer $apiKey")
            conn.setRequestProperty("Content-Type", "application/json")
            conn.connectTimeout = 30_000
            conn.readTimeout = 30_000
            conn.outputStream.use { it.write(body.toByteArray()) }
            if (conn.responseCode !in 200..299) {
                LOG.warn("OpenAI chat → HTTP ${conn.responseCode}")
                return null
            }
            conn.inputStream.bufferedReader().readText()
        } finally {
            conn.disconnect()
        }
    }
}
